base/optimized_path.cc

```diff
@@ -0,0 +1,188 @@
+std::map<std::string, OptimizedPath *> pathRegistry;
+int defaultLevel = -1;
+} // file scope
+
+void
+OptimizedPath::
+setDefault(int frequency)
+{
+    if (frequency == DEFAULT)
+        frequency = getDefaultLevel();
+    std::unique_lock<std::mutex> guard(pathRegistryMutex);
+    if (defaultLevel == frequency)
+        return;  // no change
+    defaultLevel = frequency;
+
+    // Update the short-circult flag for all those set to the default
+    for (auto & p: pathRegistry) {
+        if (p.second->itl->level == DEFAULT) {
+            p.second->alwaysTake.store(defaultLevel == ALWAYS,
+                                       std::memory_order_relaxed);
+        }
+    }
+}
+
+void
+OptimizedPath::
+setOptimization(OptimizedPath * path, int frequency)
+{
+    ExcAssert(path);
+
+    path->itl->level = frequency;
+    bool alwaysTake
+        = frequency == ALWAYS
+        || (frequency == DEFAULT && defaultLevel == ALWAYS);
+    if (path->alwaysTake.load(std::memory_order_relaxed)
+        != alwaysTake)
+        path->alwaysTake.store(alwaysTake, std::memory_order_seq_cst);
+}
+
+void
+OptimizedPath::
+setOptimization(const std::string & pathName, int frequency)
+{
+    std::unique_lock<std::mutex> guard(pathRegistryMutex);
+    auto it = pathRegistry.find(pathName);
+    if (it == pathRegistry.end())
+        throw ML::Exception("Couldn't find path name '"
+                            + pathName + "' setting optimization level");
+    setOptimization(it->second, frequency);
+}
+
+size_t
+OptimizedPath::
+setOptimizationRegex(const std::string & pathNameRegex,
+                     int frequency)
+{
+    std::regex regex(pathNameRegex);
+    std::unique_lock<std::mutex> guard(pathRegistryMutex);
+    size_t result = 0;
+    for (auto & p: pathRegistry) {
+        if (std::regex_match(p.first, regex)) {
+            ++result;
+            setOptimization(p.second, frequency);
+        }
+    }
+    return result;
+}
+
+OptimizedPath::
+OptimizedPath(const std::string & name)
+    : alwaysTake(true),
+      itl(new (storage) Itl(name), [] (Itl *) {})
+{
+    std::unique_lock<std::mutex> guard(pathRegistryMutex);
+    // Initialize here to avoid problems if this library is initialized after
+    // others.
+    if (defaultLevel == -1)
+        defaultLevel = getDefaultLevel();
+    setOptimization(this, defaultLevel);
+    if (!pathRegistry.insert({name, this}).second) {
+        throw ML::Exception("Optimized path '" + name + "' was registered twice");
+    }
+}
+
+OptimizedPath::
+~OptimizedPath()
+{
+    std::unique_lock<std::mutex> guard(pathRegistryMutex);
+    pathRegistry.erase(itl->name);
+}
+
+bool
+OptimizedPath::
+maybeTake() const
+{
+    auto level = itl->level;
+    if (level == DEFAULT)
+        level = defaultLevel;
+
+    bool result;
+    if (level == NEVER) {
+        result = false;
+    }
+    else if (level == ALWAYS) {
+        result = true;
+    }
+    else {
+        result = rand_r(&itl->randSeed) < level;
```

Comment:
If I've followed properly, this will always yield true. randSeed = 1, the only uncovered level is SOMETIMES which is max int / 2. We end up with
rand_r(1) < max int / 2